### PR TITLE
[Blazor] Prerendered state

### PR DESCRIPTION
**Pull request by javiercn**
[Blazor] Adds support for persting prerendered state on Blazor Web applications.
* Persists state both for server and webassembly as necessary.
* Initializes the state when a given interactive runtime is initialized and renders the first set of components.
  * On WebAssembly, this is the first time the app starts.
  * On Server this happens every time a circuit starts.
* The state is available during the first render, until the components reach quiescence.

The approach we follow is different for server and webassembly:
* On Server, we support initializing the circuit with an empty set of descriptors and in that case, we delay initialization until the first `UpdateRootComponents` call is issued.
  * This is because it's hard to deal with the security constraints imposed by starting a new circuit multiple times, and its easier to handle them within UpdateRootComponents. We might switch this approach in the future to go through `StartCircuit` too.
* On WebAssembly, we query for the initial set of webassembly components when we are starting the runtime in a Blazor Web Scenario.
  * We do this because Blazor WebAssembly offers a programatic API to render root components at a given location defined by their selectors, so we need to make sure that those components can receive state at the same time the initial set of WebAssembly components added to the page.

There are a set of tests validating different behaviors with regards to enhanced navigation and streaming rendering, as well as making sure that auto mode can access the state on Server and WebAssembly, and that Server gets new state every time a circuit is opened.

### CODE REVIEW

**Review by MackinnonBuck — COMMENTED** [review]
The approach looks great to me 👍

File: src/Components/Components/src/IPersistentComponentStateStore.cs [hunk 1]
```diff
@@ -20,4 +20,11 @@ public interface IPersistentComponentStateStore
     /// <param name="state">The serialized state to persist.</param>
     /// <returns>A <see cref="Task" /> that completes when the state is persisted to disk.</returns>
     Task PersistStateAsync(IReadOnlyDictionary<string, byte[]> state);
+
+    /// <summary>
+    /// Returns a value that indicates whether the store supports the given <see cref="IComponentRenderMode"/>.
+    /// </summary>
+    /// <param name="renderMode">The <see cref="IComponentRenderMode"/> in question.</param>
+    /// <returns><c>true</c> if the render mode is supported by the store, otherwise <c>false</c>.</returns>
+    bool SupportsRenderMode(IComponentRenderMode renderMode) => true;
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/IPersistentComponentStateStore.cs:29]: Adding an interface member, even with a default implementation, is technically considered a breaking change as per https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules but that doc isn't specific about the cases where it can fail. Perhaps it's fine here.

File: src/Components/Components/src/RegistrationContext.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components;
+
+internal readonly record struct RegistrationContext(Func<Task> Callback, IComponentRenderMode? RenderMode);
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RegistrationContext.cs:6]: Can you give this a more specific name? Neither "registration" nor "context" really communicate anything without further context. I know it's only internal but we still have to understand the code :) Suggestion: `PersistComponentStateRegistration`.

File: src/Components/Components/src/PersistenceContext.cs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components;
+
+internal readonly record struct PersistenceContext(IDictionary<string, byte[]> State);
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/PersistenceContext.cs:6]: Maybe rename to `PersistentComponentStateStore`? Also does it have do operate on `IDictionary` or could it be `Dictionary`?

File: src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs [hunk 4]
```diff
@@ -48,43 +46,74 @@ public async Task RestoreStateAsync(IPersistentComponentStateStore store)
...
+            if (registration.Callback.Target is IComponent component)
+            {
+                var componentRenderMode = renderer.GetComponentRenderMode(component);
+                _registeredCallbacks[i] = new RegistrationContext(registration.Callback, componentRenderMode);
+                continue;
+            }
+
+            throw new InvalidOperationException(
+                $"The registered callback {registration.Callback.Method.Name} must be associated with a component or define" +
+                $" an explicit render mode type during registration.");
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs:99]: Is this a breaking change?

File: src/Components/Components/src/PersistenceContext.cs (hunk 3 above)
Comment (javiercn) [reply src/Components/Components/src/PersistenceContext.cs:6]: I've reverted all this stuff to the old implementation, it's not needed.

File: src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs (hunk 4 above)
Comment (javiercn) [reply src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs:99]: I'm pushing an update; it's not going to be a breaking change because we are only going to do this when we can't infer the render mode. In any other case, it won't be needed. There is a test that covers that behavior.

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 5]
```diff
@@ -133,7 +133,20 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
...
         => GetRequiredComponentState(componentId);
 
-    internal ComponentState GetComponentState(IComponent component)
+    /// <summary>
+    /// Gets the <see cref="IComponentRenderMode"/> for a given component if available.
+    /// </summary>
+    /// <param name="component">The component type</param>
+    /// <returns></returns>
+    protected internal virtual IComponentRenderMode? GetComponentRenderMode(IComponent component)
+        => null;
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/Renderer.cs:142]: I'm a bit torn about this as it's possible to avoid exposing this new API (e.g., with the "filter callback" I prototyped) but at the same time I don't have a really objective reason to say it's important to avoid exposing it.

It looks like the tradeoff is:

 * With filter callback:
   * (Good) No need have a renderer-level API that returns component rendermodes
   * (Arguably good) The exact rules about resolving "registration rendermode" and "component rendermode" are delegated to the renderer, so each renderer could have its own semantics as required
   * (Bad) The API with a filter callback is less obvious
 * With `GetComponentRenderMode` and other aspects of how it's done in this PR:
   * The opposite, i.e., the API is a bit more obvious but at the cost that we need to lock all renderers to being able to say at any moment what is the rendermode for a given component, and we can not allow renders to vary anything about the semantics of how "registration rendermode" and "component rendermode" are combined.

Do you agree that's the tradeoff? I'm not trying to demand a change here since both seem like valid approaches. I just want to be sure I've understood and check if you think there are other subtleties or if it might be worth going to the "filter callback" approach (would be a pretty easy change within this PR).

File: src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs [hunk 6]
```diff
@@ -18,7 +18,7 @@ namespace Microsoft.AspNetCore.Components.Endpoints;
 internal partial class RazorComponentEndpointInvoker : IRazorComponentEndpointInvoker
 {
     private readonly EndpointHtmlRenderer _renderer;
-    private readonly ILogger<RazorComponentEndpointInvoker> _logger;
+    private readonly ILogger<RazorComponentEndpointInvoker> _logger;
```
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:21]:
```suggestion
    private readonly ILogger<RazorComponentEndpointInvoker> _logger;
```

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs [hunk 7]
```diff
@@ -33,19 +34,39 @@ protected override IComponent ResolveComponentForRenderMode([DynamicallyAccessed
...
     }
 
+    protected override IComponentRenderMode? GetComponentRenderMode(IComponent component)
+    {
+        var componentState = GetComponentState(component);
+        var ssrRenderBoundary = GetClosestRenderModeBoundary(componentState);
+
+        if (ssrRenderBoundary is null)
+        {
+            throw new InvalidOperationException("Cannot infer render mode.");
```
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs:44]: This seems like a bug. For SSR components, there will be no enclosing boundary. How come this doesn't throw in practice? Are there tests covering what happens if purely SSR components try to persist state?

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.PrerenderingState.cs [hunk 8]
```diff
@@ -15,6 +16,45 @@ internal partial class EndpointHtmlRenderer
...
+        // The persistence state manager checks if the store implements
+        // IEnumerable<IPersistentComponentStateStore> and if so, it invokes PersistStateAsync on each store
+        // for each of the render mode callbacks defined.
+        // We pass in a composite store with fake stores for each render mode that only take care of
+        // creating a copy of the state for each render mode.
+        // Then, we copy the state from the auto store to the server and webassembly stores and persist
+        // the real state for server and webassembly render modes.
+        // This makes sure that:
+        // 1. The persistence state manager is agnostic to the render modes.
+        // 2. The callbacks are run only once, even if the state ends up persisted in multiple locations.
```
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.PrerenderingState.cs:37]: That's a nice technique.

File: src/Components/Components/src/RegistrationContext.cs (hunk 2 above)
Comment (javiercn) [reply src/Components/Components/src/RegistrationContext.cs:6]: I've renamed this and changed it for a regular readonly struct, this was leftover from before

File: src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs [hunk 9]
```diff
@@ -26,12 +26,13 @@ internal class SSRRenderModeBoundary : IComponent
 
     [DynamicallyAccessedMembers(Component)]
     private readonly Type _componentType;
-    private readonly IComponentRenderMode _renderMode;
     private readonly bool _prerender;
     private RenderHandle _renderHandle;
     private IReadOnlyDictionary<string, object?>? _latestParameters;
     private string? _markerKey;
 
+    public IComponentRenderMode RenderMode { get; private set; }
```
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:34]:
```suggestion
    public IComponentRenderMode RenderMode { get; }
```

I don't think it gets written outside the constructor.

File: src/Components/Server/src/Circuits/CircuitFactory.cs [hunk 10]
```diff
@@ -74,9 +80,13 @@ internal sealed partial class CircuitFactory : ICircuitFactory
             serverComponentDeserializer,
             _loggerFactory.CreateLogger<RemoteRenderer>(),
             jsRuntime,
-            jsComponentInterop);
+            jsComponentInterop);
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Circuits/CircuitFactory.cs:83]:
```suggestion
            jsComponentInterop);
```

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 11]
```diff
@@ -130,7 +140,20 @@ public Task InitializeAsync(ProtectedPrerenderComponentApplicationStore store, C
...
+                {
+                    store.ExistingState.Clear();
+                }
+
+                // This variable is used to track that this is the first time we are updating components.
+                // In Blazor Web scenarios the app will send an initial empty list of descriptors,
+                // so we want to make sure that we allow setting up the state in that case.
+                // In Blazor Server the initial set of descriptors is provided via the call to Start, so
+                // we want to make sure we don't take any state afterwards.
+                _isFirstUpdate = Descriptors.Count == 0;
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Circuits/CircuitHost.cs:156]: What if it goes back down to zero later, after the first render?

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 12]
```diff
@@ -702,6 +725,111 @@ private async Task TryNotifyClientErrorAsync(IClientProxy client, string error,
...
+                            }
+                            break;
+                        case RootComponentOperationType.Update:
+                            var componentType = Renderer.GetExistingComponentType(operation.ComponentId.Value);
+                            if (descriptor.ComponentType != componentType)
+                            {
+                                Log.InvalidComponentTypeForUpdate(_logger, message: "Component type mismatch.");
+                            }
+
+                            _ = Renderer.UpdateRootComponentAsync(operation.ComponentId.Value, descriptor.Parameters);
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Circuits/CircuitHost.cs:798]: Why is it OK to discard the task here?

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 13]
```diff
@@ -702,6 +725,111 @@ private async Task TryNotifyClientErrorAsync(IClientProxy client, string error,
...
+                    for (var i = 0; i < operations.Length; i++)
+                    {
+                        var operation = operations[i];
+                        if (operation.Item1.Type != RootComponentOperationType.Add)
+                        {
+                            throw new InvalidOperationException($"The first set of update operations must always be of type {nameof(RootComponentOperationType.Add)}");
+                        }
+                    }
+
+                    pendingTasks = new Task[operations.Length];
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Circuits/CircuitHost.cs:775]: The use of an array, rather than `List<Task>`, is pretty surprising. There will often be gaps in the array then, since some of the operations don't add tasks. Why not use `List<Task>`, at least to avoid forcing future devs to figure out if the gaps serve some purpose?

File: src/Components/Server/src/ComponentHub.cs [hunk 14]
```diff
@@ -160,6 +160,26 @@ public async ValueTask<string> StartCircuit(string baseUri, string uri, string s
         }
     }
 
+    public async Task UpdateRootComponents(string serializedComponentOperations, string applicationState)
+    {
+        if (!_serverComponentSerializer.TryDeserializeRootComponentOperations(serializedComponentOperations, out var operations))
+        {
+            // Return error
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/ComponentHub.cs:167]: TODO?

File: src/Components/Shared/src/DefaultAntiforgeryStateProvider.cs [hunk 15]
```diff
@@ -25,7 +26,7 @@ public DefaultAntiforgeryStateProvider(PersistentComponentState state)
         {
             state.PersistAsJson(PersistenceKey, _currentToken);
             return Task.CompletedTask;
-        });
+        }, new InteractiveAutoRenderMode());
```
Comment (SteveSandersonMS) [inline src/Components/Shared/src/DefaultAntiforgeryStateProvider.cs:29]: Does this mean we're going to emit prerendered state for WebAssembly even if you're not using WebAssembly? In fact, even if you're not using any interactive rendering at all?

File: src/Components/Shared/src/RootComponentTypeCache.cs [hunk 16]
```diff
@@ -6,6 +6,7 @@
 using System.Reflection;
 
 namespace Microsoft.AspNetCore.Components;
+using static Microsoft.AspNetCore.Internal.LinkerFlags;
```
Comment (SteveSandersonMS) [inline src/Components/Shared/src/RootComponentTypeCache.cs:9]:
```suggestion
```

Since this is the only change in the file perhaps it's not needed.

File: src/Components/Components/src/RenderTree/Renderer.cs (hunk 5 above)
Comment (javiercn) [reply src/Components/Components/src/RenderTree/Renderer.cs:142]: > The opposite, i.e., the API is a bit more obvious but at the cost that we need to lock all renderers to being able to say at any moment what is the rendermode for a given component, and we can not allow renders to vary anything about the semantics of how "registration rendermode" and "component rendermode" are combined.

It's not mandatory for a renderer to provide a RenderMode, in fact, the endpoint HTML renderer is the only one that provide is it.

We could remove that from the public API of the renderer and instead pass it as a callback to `PersistStateAsync` but I didn't want to add an additional parameter to it as that required a new overload and potential breaking change.

From my view, you only implement `GetComponentRenderMode` if your renderer handles multiple render modes. Otherwise, the default implementation is fine. As in any case, any caller will have to account for the component not having a render mode.

Alternatively, we could also put this API in `ComponentState` which would mean something like `Renderer.GetComponentState(component).GetRenderMode()`

Comment (javiercn) [reply src/Components/Components/src/RenderTree/Renderer.cs:142]: IComponentRenderMode permeates through the core `Microsoft.AspNetCore.Components` API, so I think it's fine having methods that interact with the abstraction as long as we don't bake in behaviors specific to the concrete set of render modes we provide.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs [hunk 17]
```diff
@@ -151,17 +151,28 @@ internal async Task RunAsyncCore(CancellationToken cancellationToken, WebAssembl
...
             {
                 var (rootComponents, renderer, initializationTcs) = state;
-
                 try
                 {
                     // Here, we add each root component but don't await the returned tasks so that the
                     // components can be processed in parallel.
                     var count = rootComponents.Count;
-                    var pendingRenders = new Task[count];
+                    var pendingRenders = new Task[count + mappings.Length];
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs:175]: Could we make it a `List<Task>` instead so we don't have to worry about the capacity, or the indices having special meanings?

File: src/Components/test/E2ETest/Tests/StatePersistenceTest.cs [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+        Browser.Click(By.Id("page-no-components-link"));
+        // Ensure that the circuit is gone.
+        await Task.Delay(1000);
+        Browser.Click(By.Id("page-with-components-link-and-state"));
+        RenderComponentsWithPersistentStateAndValidate(suppresEnhancedNavigation: false, mode, typeof(InteractiveServerRenderMode), streaming, stateValue: "other");
+    }
+
+    private void BlockWebAssemblyResourceLoad()
+    {
+        ((IJavaScriptExecutor)Browser).ExecuteScript("sessionStorage.setItem('block-load-boot-resource', 'true')");
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Tests/StatePersistenceTest.cs:136]: What is this used for? I can't see anything that reads the value.

File: src/Components/test/testassets/Components.TestServer/Components - Backup.TestServer.csproj [hunk 19]
```diff
@@ -0,0 +1,46 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/Components - Backup.TestServer.csproj:1]: To be removed, I assume.

File: src/Components/test/testassets/Components.TestServer/Components.TestServer.csproj [hunk 20]
```diff
@@ -42,5 +42,5 @@
       <_Parameter2 Condition="'$(IsHelixJob)' == 'true'">..\BasicTestApp</_Parameter2>
     </AssemblyAttribute>
   </ItemGroup>
-
+
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/Components.TestServer.csproj:45]: Revert

File: src/Components/test/testassets/Components.TestServer/Properties/launchSettings.json [hunk 21]
```diff
@@ -11,6 +11,7 @@
         // time you restart the test server
         "TESTSERVER_USE_DETERMINISTIC_PORTS": "true"
       },
+      //"inspectUri": "{wsProtocol}://{url.hostname}:5005/_framework/debug/ws-proxy?browser={browserInspectUri}",
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/Properties/launchSettings.json:14]: Revert

File: src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs [hunk 22]
```diff
@@ -47,6 +47,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
         if (env.IsDevelopment())
         {
             app.UseDeveloperExceptionPage();
+            app.UseWebAssemblyDebugging();
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:50]: Do we need this?

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/PersistStateComponents.razor [hunk 23]
```diff
@@ -0,0 +1,32 @@
+﻿@page "/persist-state"
+
+<h1>Persist State Components</h1>
+
+@if (Server.HasValue && Server.Value) {
+    <strong>Server Persist State Component</strong>
+    <TestContentPackage.PersistStateComponent KeyName="server" @rendermode="@(new InteractiveServerRenderMode(true))" />
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/PersistStateComponents.razor:7]:
```suggestion
    <TestContentPackage.PersistStateComponent KeyName="server" @rendermode="@RenderModes.InteractiveServer" />
```

etc in other cases below

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/PersistentState/EndStreamingPage.razor [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+    [SupplyParameterFromQuery(Name = "streaming-id")] public string StreamingId { get; set; }
+
+    [Inject] public AsyncOperationService StreamingManager { get; set; }
+
+    protected override void OnInitialized()
+    {
+        if (string.IsNullOrEmpty(StreamingId))
+        {
+            throw new InvalidOperationException("StreamingId is required.");
+        }else
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/PersistentState/EndStreamingPage.razor:30]:
```suggestion
        }
        else
```

File: src/Components/test/testassets/TestContentPackage/PersistentComponents/NonStreamingComponentWithPersistentState.razor [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            _stateValue = "fresh";
+        }
+
+        if (RunningOnServer)
+        {
+            _interactiveRuntime = "none";
+            _stateFound = true;
+            _stateValue = ServerState ?? "restored";
+        }else
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestContentPackage/PersistentComponents/NonStreamingComponentWithPersistentState.razor:37]:
```suggestion
        }
        else
```

File: src/Components/Components/src/PublicAPI.Unshipped.txt [hunk 26]
```diff
@@ -44,6 +45,7 @@ Microsoft.AspNetCore.Components.RenderTree.NamedEventChangeType
 Microsoft.AspNetCore.Components.RenderTree.NamedEventChangeType.Added = 0 -> Microsoft.AspNetCore.Components.RenderTree.NamedEventChangeType
 Microsoft.AspNetCore.Components.RenderTree.NamedEventChangeType.Removed = 1 -> Microsoft.AspNetCore.Components.RenderTree.NamedEventChangeType
 Microsoft.AspNetCore.Components.RenderTree.RenderBatch.NamedEventChanges.get -> Microsoft.AspNetCore.Components.RenderTree.ArrayRange<Microsoft.AspNetCore.Components.RenderTree.NamedEventChange>?
+Microsoft.AspNetCore.Components.RenderTree.Renderer.GetComponentState(Microsoft.AspNetCore.Components.IComponent! component) -> Microsoft.AspNetCore.Components.Rendering.ComponentState!
```
Comment (javiercn) [inline src/Components/Components/src/PublicAPI.Unshipped.txt:48]: Todo: Figure out why this change shows up here.

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 27]
```diff
@@ -133,7 +133,20 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
...
+
+    /// <summary>
+    ///
+    /// </summary>
+    /// <param name="component"></param>
+    /// <returns></returns>
+    protected internal ComponentState GetComponentState(IComponent component)
```
Comment (javiercn) [inline src/Components/Components/src/RenderTree/Renderer.cs:149]: `EndpointHtmlRenderer` needs this API to find the component state from a callback target.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs [hunk 28]
```diff
@@ -33,19 +34,39 @@ protected override IComponent ResolveComponentForRenderMode([DynamicallyAccessed
...
 
+    protected override IComponentRenderMode? GetComponentRenderMode(IComponent component)
+    {
+        var componentState = GetComponentState(component);
+        var ssrRenderBoundary = GetClosestRenderModeBoundary(componentState);
+
+        if (ssrRenderBoundary is null)
+        {
+            return null;
+        }
```
Comment (javiercn) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs:45]: This threw an exception; I am not sure that's the right behavior. I've switched it to return null, however it feels that we are missing a way to represent a `StaticRenderMode`

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 11 above)
Comment (javiercn) [reply src/Components/Server/src/Circuits/CircuitHost.cs:156]: It can't. The descriptors are populated at the time the circuit is created and no longer updated (it's a readonly list). That's essentially what I am using to distinguish between old blazor server apps and blazor web.

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 13 above)
Comment (javiercn) [reply src/Components/Server/src/Circuits/CircuitHost.cs:775]: We are on the first update, and we just validated that all operations are `Add`, so we know the exact number of tasks we need to capture and there won't be any gaps.

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 12 above)
Comment (javiercn) [reply src/Components/Server/src/Circuits/CircuitHost.cs:798]: Any exception that happens will be dispatched through the UnhandledExceptionHandler in the remote renderer

File: src/Components/Server/src/Circuits/ServerComponentDeserializer.cs [hunk 29]
```diff
@@ -270,6 +270,74 @@ private bool IsWellFormedServerComponent(ComponentMarker record)
         return (componentDescriptor, serverComponent);
     }
 
+    public bool TryDeserializeRootComponentOperations(string serializedComponentOperations, out (RootComponentOperation, ComponentDescriptor?)[] operations)
```
Comment (javiercn) [inline src/Components/Server/src/Circuits/ServerComponentDeserializer.cs:273]: We should check for the uniqueness of the Update/Delete operations to ensure that there is at most one update/delete per component ID.

File: src/Components/Shared/src/DefaultAntiforgeryStateProvider.cs (hunk 15 above)
Comment (javiercn) [reply src/Components/Shared/src/DefaultAntiforgeryStateProvider.cs:29]: No, it's covered in a later commit, but essentially if you only do `AddWebAssemblyRenderMode` or `AddServerRenderMode` we only emit the comment for webassembly or server respectively. 

If you have declared that you are using both, then we will emit it. I haven't done it, but we can look at the set of actual render modes that a component used during a request and narrow down to a single render mode if possible too, in that situation.

The point here, is that auto means, "wherever it's actually needed" as opposed to all locations.

File: src/Components/test/E2ETest/Tests/StatePersistenceTest.cs (hunk 18 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Tests/StatePersistenceTest.cs:136]: It forces the app to start on Blazor Server when auto is enabled on the page. Other similar tests follow this approach.

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 11 above)
Comment (javiercn) [reply src/Components/Server/src/Circuits/CircuitHost.cs:156]: Once you provide an initial set of descriptors via a call to "StartCircuit" you are locked on to "Blazor Server" behaviors and vice versa.

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 12 above)
Comment (SteveSandersonMS) [reply src/Components/Server/src/Circuits/CircuitHost.cs:798]: OK, could you add a comment to that effect?

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 13 above)
Comment (SteveSandersonMS) [reply src/Components/Server/src/Circuits/CircuitHost.cs:775]: Oh right, I see. It would be more obvious if the variable name was `pendingTasksForFirstRender` rather than just `pendingTasks`, so it's clearer the code below behaves how it does due to the assumptions in the code above.

**Review by SteveSandersonMS — APPROVED** [review]
Excellent! I understand there are still some tests to add but we've been through this in enough detail now so I'm fine with approving. Please let me know if anything meaningful still has to change about the runtime code.

**Comment by marinasundstrom** [comment]
I'm running the test host, but I can't figure out how to navigate to the page that is testing this. All I get is ``/subdir``, and there it is not included. What do I need to do?
